pydy/codegen/ode_function_generator.py

```diff
@@ -0,0 +1,705 @@
+            # is a dictionary which at least contains the key 'constants'.
+            # So if the last arg is so, then extract.
+            args[-1]['constants']
+        # ValueError is needed for older NumPy versions.
+        except (KeyError, IndexError, ValueError):
+            return args
+        else:
+            new_args = list(args[:-1])  # gets x and t
+
+            if self.specifieds is not None:
+                new_args.append(args[-1]['specified'])
+
+            new_args.append(args[-1]['constants'])
+
+            return tuple(new_args)
+
+    def _parse_specifieds(self, x, t, r, p):
+
+        if isinstance(r, dict):
+            # TODO : This should be instatiated outside of the RHS function.
+            specified_values = np.zeros(self.num_specifieds)
+
+            for k, v in r.items():
+                # TODO : Not sure if this is the best check here.
+                if isinstance(type(k), UndefinedFunction):
+                    k = (k,)
+                idx = [self.specifieds.index(symmy) for symmy in k]
+                try:
+                    specified_values[idx] = v(x, t)
+                except TypeError:  # not callable
+                    # If not callable, then it should be a float, ndarray,
+                    # or indexable.
+                    specified_values[idx] = v
+        else:
+            # More efficient.
+            try:
+                specified_values = r(x, t)
+            except TypeError:  # not callable.
+                # If not callable, then it should be a float or ndarray.
+                specified_values = r
+            # If the value is just a float, then convert to a 1D array.
+            try:
+                len(specified_values)
+            except TypeError:
+                specified_values = np.asarray([specified_values])
+
+        return x, t, specified_values, p
+
+
+class FullRHSMixin(MixinBase):
+
+    def create_rhs_function(self):
+        """Returns a function in the form expected by scipy.integrate.odeint
+        that computes the derivatives of the states."""
+
+        def rhs(*args):
+            # args: x, t, p
+            # args: x, t, r, p
+
+            args = self._parse_old_style_extra_args(*args)
+
+            args = self._parse_constants(*args)
+
+            if self.specifieds is not None:
+                args = self._parse_specifieds(*args)
+
+            q = args[0][:self.num_coordinates]
+            u = args[0][self.num_coordinates:]
+
+            xdot = self.eval_arrays(q, u, *args[2:])
+
+            return xdot
+
+        rhs.__doc__ = self._generate_rhs_docstring()
+
+        return rhs
+
+
+class FullMassMatrixMixin(MixinBase):
+
+    def create_rhs_function(self):
+
+        def rhs(*args):
+            # args: x, t, p
+            # args: x, t, r, p
+
+            args = self._parse_old_style_extra_args(*args)
+
+            args = self._parse_constants(*args)
+
+            if self.specifieds is not None:
```

Comment:
Yeah, I don't like how I've defined everything here so much. It also takes way too much time to run these parsing functions to support all the different args in each rhs eval. I'll clean it up.